test/main.go

```diff
@@ -0,0 +1,149 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	appsv1 "k8s.io/client-go/kubernetes/typed/apps/v1"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func main() {
+	kubeconfig := filepath.Join(
+		os.Getenv("HOME"), ".kube", "config",
+	)
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		log.Fatal(err)
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Fatal(err)
+	}
+	// NOTE: Certain features are in different clientsets. For example,
+	// Pods would be in CoreV1, not AppsV1
+	api := clientset.AppsV1()
+
+	// create rbac
+	kubectlWrapper("create", "deploy/rbac.yaml")
+	fmt.Println("Created rbac")
+
+	// create crd
+	kubectlWrapper("create", "deploy/crd.yaml")
+	fmt.Println("Created crd")
+
+	// create operator
+	kubectlWrapper("create", "deploy/operator.yaml")
+	fmt.Println("Created operator")
+
+	deploymentReplicaCheck(api, "memcached-operator", 1, 60)
+
+	// create example-memcached yaml file
+	file, err := os.OpenFile("deploy/cr.yaml", os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer file.Close()
+
+	_, err = file.WriteString("apiVersion: \"cache.example.com/v1alpha1\"\nkind: \"Memcached\"\nmetadata:\n  name: \"example-memcached\"\nspec:\n  size: 3")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	file.Close()
+
+	// apply/create example-memcached deployment
+	kubectlWrapper("apply", "deploy/cr.yaml")
+
+	deploymentReplicaCheck(api, "example-memcached", 3, 60)
+
+	// update deployment to 4 replicas
+	cr, err := ioutil.ReadFile("deploy/cr.yaml")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	newCr := bytes.Replace(cr, []byte("size: 3"), []byte("size: 4"), -1)
+
+	file, err = os.OpenFile("deploy/cr.yaml", os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer file.Close()
+
+	_, err = file.Write(newCr)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	file.Close()
+
+	// apply updated example-memcached deployment
+	kubectlWrapper("apply", "deploy/cr.yaml")
+
+	deploymentReplicaCheck(api, "example-memcached", 4, 60)
+}
+
+func printDeployments(deployments *v1.DeploymentList) {
+	template := "%-40s%-10s\n"
+	fmt.Printf(template, "NAME", "NUM_REPLICAS")
+	for _, deployment := range deployments.Items {
+		fmt.Printf(
+			template,
+			deployment.Name,
+			strconv.Itoa(int(deployment.Status.AvailableReplicas)),
+		)
+	}
+}
+
+func deploymentReplicaCheck(api appsv1.AppsV1Interface, name string, replicas, timeout int) {
+	count := 0
+
+	// setup list options
+	listOptions := metav1.ListOptions{
+		LabelSelector: "",
+		FieldSelector: "",
+	}
+
+outerloop:
```

Comment:
We can probably make the retry-until-true-or-timeout check into a util and use that for all of our checks.
See how the etcd-operator does it:
https://github.com/coreos/etcd-operator/blob/master/pkg/util/retryutil/retry_util.go
https://github.com/coreos/etcd-operator/blob/master/test/e2e/e2eutil/wait_util.go#L255-L277

We can add a similar retry-timeout utility to the sdk.